### ISSUE TITLE
splog's performance in 1.10.0 with SPDLOG_USE_STD_FORMAT

### ISSUE THREAD

**Issue by USAFrenzy** [gabime/spdlog#2344]
When using the new release of spdlog and building with the option ```SPDLOG_USE_STD_FORMAT```, the formatting times are heavily slowing down the output of this library. Compiling with the ```/utf-8``` compiler flag as is suggested by Microsoft speeds up the formatting times, but the performance is still heavily dampened when compared to the fmtlib usage.  What I've noticed is that using ```std::vformat_to()``` and explicitly specializing the context used speeds this up dramatically versus using MSVC's version of ```std::vformat()```. So with this being said, I was wondering what your thoughts on changing the lines pertaining to the following were?
``` 
memory_buf_t buf = std::vformat(fmt, std::make_format_args(std::forward<Args...>(args)...));

``` 
and using a similar macro based on ```_MSC_VER```
that looks like:
```
#    if _MSC_VER >= 1930
#        define STD_CONTEXT std::back_insert_iterator<std::basic_string<char>>
#        define STD_VFORMAT_TO(container, message, ...)                                                                                    \
            std::vformat_to<STD_CONTEXT>(std::back_inserter(container), message, std::make_format_args(__VA_ARGS__))
#    elif _MSC_VER >= 1929
#        define STD_CONTEXT std::basic_format_context<std::back_insert_iterator<std::basic_string<char>>, char>
#        define STD_VFORMAT_TO(container, message, ...)                                                                                    \
            std::vformat_to(std::back_inserter(container), message, std::make_format_args<STD_CONTEXT>(__VA_ARGS__))
#    else
#        define STD_VFORMAT_TO(container, message, ...) container = std::vformat(message, std::make_format_args(__VA_ARGS__));
#    endif
```
to 
```
memory_buf_t buf;
STD_VFORMAT_TO(buf, fmt, std::forward<Args>(args)...);
```

with context and place of that context changing only due to the apparent difference in implementation from VS v16 and  VS v17.
The performance increased from ~3.88 microseconds to ~2.37 microseconds on my machine and if you add a way to store the current locale and just pass it to the macro for the version of ```std::vformat_to()``` that takes a locale, the performance is increased a bit more (I haven't tested this aspect due to not messing around with spdlog enough to try that change). Just as a reference, using spdlog with fmtlib on my machine generally runs with the output being around an average of ~1.08 microseconds, hence the issue. This could also be user error on my part as I may not be aware of a setting spdlog uses in this case, but I just wanted to bring this up if it wasn't already known.

EDIT: Edited to provide fallback to current usage when used on other platforms instead
EDIT 2: Sorry about the second edit, realized the typo in the #elif statement

**Comment by USAFrenzy** [comment]
Using the bench.cpp provided by spdlog, the changes are reflected below:

Original Version:
```
[info] **************************************************************
[info] Single threaded: 250000 messages
[info] **************************************************************
[info] basic_st                       Elapsed: 0.15 secs        1,642,688/sec
[info] basic_st/backtrace-on          Elapsed: 0.15 secs        1,622,357/sec
[info]
[info] rotating_st                    Elapsed: 0.17 secs        1,510,562/sec
[info] rotating_st/backtrace-on       Elapsed: 0.21 secs        1,208,833/sec
[info]
[info] daily_st                       Elapsed: 0.16 secs        1,599,849/sec
[info] daily_st/backtrace-on          Elapsed: 0.20 secs        1,270,888/sec
[info]
[info] level-off                      Elapsed: 0.00 secs      137,597,005/sec
[info] level-off/backtrace-on         Elapsed: 0.10 secs        2,534,774/sec
[info] **************************************************************
[info] Multi threaded: 1 threads, 250,000 messages
[info] **************************************************************
[info] basic_mt                       Elapsed: 0.17 secs        1,511,001/sec
[info] basic_mt/backtrace-on          Elapsed: 0.19 secs        1,328,653/sec
[info]
[info] rotating_mt                    Elapsed: 0.18 secs        1,404,286/sec
[info] rotating_mt/backtrace-on       Elapsed: 0.19 secs        1,297,037/sec
[info]
[info] daily_mt                       Elapsed: 0.16 secs        1,560,843/sec
[info] daily_mt/backtrace-on          Elapsed: 0.21 secs        1,212,990/sec
[info]
[info] level-off                      Elapsed: 0.00 secs      137,559,150/sec
[info] level-off/backtrace-on         Elapsed: 0.10 secs        2,521,554/sec
[info] **************************************************************
[info] Multi threaded: 4 threads, 250,000 messages
[info] **************************************************************
[info] basic_mt                       Elapsed: 0.23 secs        1,097,544/sec
[info] basic_mt/backtrace-on          Elapsed: 0.30 secs          834,873/sec
[info]
[info] rotating_mt                    Elapsed: 0.25 secs        1,015,892/sec
[info] rotating_mt/backtrace-on       Elapsed: 0.33 secs          756,162/sec
[info]
[info] daily_mt                       Elapsed: 0.25 secs        1,001,905/sec
[info] daily_mt/backtrace-on          Elapsed: 0.33 secs          747,273/sec
[info]
[info] level-off                      Elapsed: 0.00 secs      137,234,451/sec
[info] level-off/backtrace-on         Elapsed: 0.10 secs        2,513,244/sec
```

Macro Version
```
[info] **************************************************************
[info] Single threaded: 250000 messages
[info] **************************************************************
[info] basic_st                       Elapsed: 0.13 secs        1,955,717/sec
[info] basic_st/backtrace-on          Elapsed: 0.13 secs        1,996,600/sec
[info]
[info] rotating_st                    Elapsed: 0.13 secs        1,904,892/sec
[info] rotating_st/backtrace-on       Elapsed: 0.17 secs        1,444,110/sec
[info]
[info] daily_st                       Elapsed: 0.12 secs        2,001,002/sec
[info] daily_st/backtrace-on          Elapsed: 0.18 secs        1,385,320/sec
[info]
[info] level-off                      Elapsed: 0.00 secs      145,205,320/sec
[info] level-off/backtrace-on         Elapsed: 0.07 secs        3,550,853/sec
[info] **************************************************************
[info] Multi threaded: 1 threads, 250,000 messages
[info] **************************************************************
[info] basic_mt                       Elapsed: 0.13 secs        1,960,573/sec
[info] basic_mt/backtrace-on          Elapsed: 0.18 secs        1,389,073/sec
[info]
[info] rotating_mt                    Elapsed: 0.14 secs        1,743,826/sec
[info] rotating_mt/backtrace-on       Elapsed: 0.18 secs        1,352,144/sec
[info]
[info] daily_mt                       Elapsed: 0.13 secs        1,888,892/sec
[info] daily_mt/backtrace-on          Elapsed: 0.18 secs        1,409,737/sec
[info]
[info] level-off                      Elapsed: 0.00 secs      145,509,574/sec
[info] level-off/backtrace-on         Elapsed: 0.07 secs        3,394,299/sec
[info] **************************************************************
[info] Multi threaded: 4 threads, 250,000 messages
[info] **************************************************************
[info] basic_mt                       Elapsed: 0.17 secs        1,448,919/sec
[info] basic_mt/backtrace-on          Elapsed: 0.30 secs          823,046/sec
[info]
[info] rotating_mt                    Elapsed: 0.19 secs        1,317,471/sec
[info] rotating_mt/backtrace-on       Elapsed: 0.34 secs          740,780/sec
[info]
[info] daily_mt                       Elapsed: 0.19 secs        1,293,369/sec
[info] daily_mt/backtrace-on          Elapsed: 0.32 secs          785,805/sec
[info]
[info] level-off                      Elapsed: 0.00 secs      145,611,276/sec
[info] level-off/backtrace-on         Elapsed: 0.07 secs        3,577,377/sec
```

**Comment by gabime** [comment]
I am little confused, sinice was this discussed in the pr #2170, with the opposite conclusions. Please see @sylveon comments.

**Comment by sylveon** [comment]
I was previously able to consistently see std::vformat benchmark faster, but cannot get it to do that anymore. Something must have changed since I last benchmarked it. I would not be opposed to universally replacing vformat by vformat_to, instead of a macro-based solution, that way we might be able to get rid of the `#ifdef SPDLOG_USE_STD_FORMAT` around these vformat calls and use `spdlog::fmt_lib` instead. Would support `SPDLOG_WCHAR_TO_UTF8_SUPPORT` as well (which the proposed macro does not).

**Comment by sylveon** [comment]
> I believe the reason why side-stepping the call to std::vformat() in favor of std::vformat_to() here is slightly faster is that we're explicitly stating the type of std::back_inserter to use that happens to line up with the underlying type of memory_buf_t -> removing the need to reserve an estimated size on a local function string object and creating a back_inserter context. Instead, if given an explicit back_inserter context, the std::vformat_to() seems to explicitly use this and parses the string to format into the output buffer.

I'm not quite sure I get what this means. Looking at what I see here, I have vformat and vformat to as the following:
```cpp
template <int = 0> // improves throughput, see GH-2329
_NODISCARD string vformat(const string_view _Fmt, const format_args _Args) {
    string _Str;
    _Str.reserve(_Fmt.size() + _Args._Estimate_required_capacity());
    _STD vformat_to(back_insert_iterator{_Str}, _Fmt, _Args);
    return _Str;
}

template <output_iterator<const char&> _OutputIt>
_OutputIt vformat_to(_OutputIt _Out, const string_view _Fmt, const format_args _Args) {
    if constexpr (is_same_v<_OutputIt, _Fmt_it>) {
        _Format_handler<char> _Handler(_Out, _Fmt, _Args);
        _Parse_format_string(_Fmt, _Handler);
        return _Out;
    } else {
        _Fmt_iterator_buffer<_OutputIt, char> _Buf(_STD move(_Out));
        _Format_handler<char> _Handler(_Fmt_it{_Buf}, _Fmt, _Args);
        _Parse_format_string(_Fmt, _Handler);
        return _Buf._Out();
    }
}
```

vformat's implementation is quite literally the same as what spdlog is doing here (besides the reserve to reduce memory allocations - the reservation is an optimization that should be there to reduce the amount of string resizes occuring during formatting). There is no difference between calling back_inserter and explicitly constructing a back_insert_iterator, as shown by looking at the implementation of back_inserter:
```cpp
template <class _Container>
_NODISCARD _CONSTEXPR20 back_insert_iterator<_Container> back_inserter(_Container& _Cont) noexcept /* strengthened */ {
    // return a back_insert_iterator
    return back_insert_iterator<_Container>(_Cont);
}
```

vformat_to will also directly format into the string object allocated by vformat (which is then returned by value).

My best guess as what is happening is that we've gone to the point of microbenchmarking, and either the reserve is more hurtful than benefitial for the string sizes we're dealing with, or the return by value of std::string is what's hurting (which is possible, since the entire SSO buffer needs to be copied, even if unused). It seems to be a combination of both, as increasing the string size to almost 3000 characters in my own benchmark made the difference between vformat_to and vformat in the order of 1% (but vformat_to still winning by a slight margin).

I can submit a PR to resolve this tonight.

**Comment by USAFrenzy** [comment]
@sylveon  That actually makes a lot of sense. I was under the impression that the allocation of the function local string  in regards to this was what the cause of the slow down given vformat_to() avoids the extra allocation -> the part of the back_inserter context was just me trying to reason why that difference may have been a thing since in the debugger, the function that was called when I ran this was
```cpp
_NODISCARD inline string vformat(const string_view _Fmt, const format_args _Args) {
    string _Str;
    _Str.reserve(_Fmt.size() + _Args._Estimate_required_capacity());
    _Fmt_iterator_buffer<back_insert_iterator<string>, char> _Buf(back_insert_iterator{_Str});
    _STD vformat_to(_Fmt_it{_Buf}, _Fmt, _Args);
    return _Str;
}
```
versus 
```cpp
template <output_iterator<const char&> _OutputIt>
_OutputIt vformat_to(
    _OutputIt _Out, const string_view _Fmt, const format_args_t<type_identity_t<_OutputIt>, char> _Args) {
    _Format_handler<_OutputIt, char, basic_format_context<_OutputIt, char>> _Handler(_STD move(_Out), _Fmt, _Args);
    _Parse_format_string(_Fmt, _Handler);
    return _Handler._Ctx.out();
}
```
and my thought was that explicitly stating the back_insert_iterator type when called also eliminated the construction of the back_insert_iterator. I whole-heartedly believe you're correct though and appreciate the clarification. To be clear, this would be my first insight into contributing so I'm taking this as a learning step lol

**Comment by gabime** [comment]
> I can submit a PR to resolve this tonight

Resolve by using `vformat_to` instead of `vformat` ?

**Comment by sylveon** [comment]
Yes

**Comment by gabime** [comment]
Please do.

**Comment by sylveon** [comment]
Should be addressed by #2346

**Comment by USAFrenzy** [comment]
Closing issue in lieu of the merge - thank you to the both of you!
